changeset/changeset.go

```diff
@@ -9,88 +10,101 @@ import (
+				return !c.touched[k]
+			}
+		}
+	}
+	return true
 }
 
-// AsStruct returns the changeset as a struct or an error if the data could not be decoded into the struct.
+// AsStruct returns the changeset decoded into a struct or an error if there was a problem decoding.
 func (c *Changeset) AsStruct() (any, error) {
 	s := c.Struct
-	err := c.config.decoder.Decode(s, c.Values)
+	err := c.config.Decoder.Decode(s, c.Values)
 	return s, err
 }
 
-// NewChangeset returns a new Changeset based on the old data, new data, and action
-// Typically this is called to initialize a changeset. If action is empty, the changeset
-// will always return true for Valid. Passing a non-empty action will cause the
-// changeset to make the validation errors available if the struct is not valid.
-func (cc *Config) NewChangeset(old, new url.Values, action string, obj any) *Changeset {
-	c := &Changeset{
-		Action:     action,
-		Valid:      action == "",
-		Errors:     make(map[string]any),
-		Changes:    make(map[string]any),
-		Touched:    make(map[string]bool),
-		Values:     url.Values{},
-		Struct:     obj,                               // TODO check is pointer to struct
-		StructType: reflect.TypeOf(obj).Elem().Name(), // TODO better way?
-		config:     cc,
+// NewChangeset returns a new Changeset based on the provided pointer to a struct.
+func (cc *Config) NewChangeset(obj any) (*Changeset, error) {
+	// we need a pointer to a struct to decode into
+	if reflect.TypeOf(obj).Kind() != reflect.Ptr {
+		return nil, errors.New("changeset: obj must be pointer to struct")
 	}
-	// initialize to old data
-	for k, v := range old {
-		c.Values[k] = v
+
+	c := &Changeset{
+		Struct: obj,
+		config: cc,
 	}
-	// update to new data
-	c.Update(new, action)
-	return c
+	return c, nil
 }
 
 // Update updates the changeset with new data and action. If action is empty, the changeset
-// will always return true for Valid. Passing a non-empty action will cause the
-// changeset to make the validation errors available if the struct is not valid.
-func (c *Changeset) Update(newData url.Values, action string) {
-	c.Action = action
-	c.Valid = action == ""          // default to true if no action, otherwise false
-	c.Errors = make(map[string]any) // TODO: once clear is in the language, use it here
+// will always return true for Valid(). Passing a non-empty action will cause the
+// changeset to run validations which may change the result of Valid() depending on
+// whether or not there are errors and whether or not the field was touched.
+func (c *Changeset) Update(newData url.Values, action string) error {
+	c.action = action
+	// TODO should we call Reset() if newData is nil and action is empty?
 
 	// merge old and new data and calculate changes
+	if c.Values == nil {
+		c.Values = url.Values{}
+	}
 	for k, v := range newData {
 		if !slices.Equal(c.Values[k], v) {
+			if c.Changes == nil {
+				c.Changes = make(map[string]any)
+			}
 			c.Changes[k] = v
 		}
 		c.Values[k] = v
```

Comment:
Is the old data preserved anywhere? In ecto you can ask for the old values.

Reply:
I suppose we lost the concept of "old" in the PR.  `NewChangeset` doesn't take `old` anymore which I think means `Changes` will always be equal to `Values`.  Hmmm.  Do we add them back to `NewChangeset`?

Reply:
For `NewChangeset`, it seems like `old` is the zero values.

Reply:
true if the form is initially empty but if there is already data, we currently don't track changes from there.

Reply:
I thought what would happen then it you'd call `NewChangeset`, then call `Update`, and then call `Update` again.

Or is the `old` value the "initial" value, not the value since the most recent change?

In that case, I'd suggest that `Reset` take a value to use as the base value. Then you'd do something like `NewChangeset; Reset; Update; Update; Update; ...; maybe put into and then take out of a sync.Pool; ... Reset; Update; Update; Update...`